File: public/java/040-invert-binary-tree.java
/*
 * LeetCode Problem #226: Invert Binary Tree
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/invert-binary-tree/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // APPROACH 1: Recursive DFS (Post-order) | O(n) time | O(h) space
    // EXPLAIN: Recursively invert left and right subtrees, then swap the children of the current node.
    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        TreeNode left  = invertTree(root.left);
        TreeNode right = invertTree(root.right);
        root.left  = right;
        root.right = left;
        return root;
    }

    // APPROACH 2: Recursive DFS (Pre-order) | O(n) time | O(h) space
    // EXPLAIN: Swap children first, then recurse into the already-swapped subtrees.
    public TreeNode invertTreePreorder(TreeNode root) {
        if (root == null) return null;
        TreeNode tmp  = root.left;
        root.left  = root.right;
        root.right = tmp;
        invertTreePreorder(root.left);
        invertTreePreorder(root.right);
        return root;
    }

    // APPROACH 3: Iterative BFS | O(n) time | O(w) space
    // EXPLAIN: BFS with a queue; swap children of each dequeued node, then enqueue non-null children.
    public TreeNode invertTreeBFS(TreeNode root) {
        if (root == null) return null;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            TreeNode tmp  = node.left;
            node.left  = node.right;
            node.right = tmp;
            if (node.left  != null) q.offer(node.left);
            if (node.right != null) q.offer(node.right);
        }
        return root;
    }

    // APPROACH 4: Iterative DFS (Stack) | O(n) time | O(h) space
    // EXPLAIN: Use an explicit stack instead of the call stack; same logic as pre-order recursive.
    public TreeNode invertTreeStack(TreeNode root) {
        if (root == null) return null;
        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            TreeNode tmp  = node.left;
            node.left  = node.right;
            node.right = tmp;
            if (node.left  != null) stack.push(node.left);
            if (node.right != null) stack.push(node.right);
        }
        return root;
    }

    // APPROACH 5: Morris Traversal | O(n) time | O(1) space
    // EXPLAIN: Threaded-tree traversal; swap children when visiting each node without extra memory.
    public TreeNode invertTreeMorris(TreeNode root) {
        TreeNode curr = root;
        while (curr != null) {
            if (curr.left == null) {
                TreeNode tmp = curr.left; curr.left = curr.right; curr.right = tmp;
                curr = curr.left;
            } else {
                TreeNode pred = curr.left;
                while (pred.right != null && pred.right != curr) pred = pred.right;
                if (pred.right == null) {
                    pred.right = curr;
                    curr = curr.left;
                } else {
                    pred.right = null;
                    TreeNode tmp = curr.left; curr.left = curr.right; curr.right = tmp;
                    curr = curr.left;
                }
            }
        }
        return root;
    }
}

// Made with Bob

File: public/python/040-invert-binary-tree.py
# LeetCode Problem #226: Invert Binary Tree
# Difficulty: Easy
# Link: https://leetcode.com/problems/invert-binary-tree/

from typing import Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Recursive DFS (Post-order) | O(n) time | O(h) space
# EXPLAIN: Recursively invert both subtrees, then swap the children.
# WHEN: Cleanest solution; standard recursive approach.

def invertTree_postorder(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return None
    left  = invertTree_postorder(root.left)
    right = invertTree_postorder(root.right)
    root.left  = right
    root.right = left
    return root


# ─────────────────────────────────────────────
# APPROACH 2: Recursive DFS (Pre-order) | O(n) time | O(h) space
# EXPLAIN: Swap children first, then recurse into the (already-swapped) subtrees.
# WHEN: Equivalent to post-order; some find pre-order more intuitive.

def invertTree_preorder(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return None
    root.left, root.right = root.right, root.left
    invertTree_preorder(root.left)
    invertTree_preorder(root.right)
    return root


# ─────────────────────────────────────────────
# APPROACH 3: Iterative BFS | O(n) time | O(w) space
# EXPLAIN: Level-order BFS with a deque; swap children of each dequeued node.
# WHEN: Avoids recursion; processes the tree level by level.

def invertTree_bfs(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return None
    q = deque([root])
    while q:
        node = q.popleft()
        node.left, node.right = node.right, node.left
        if node.left:
            q.append(node.left)
        if node.right:
            q.append(node.right)
    return root


# ─────────────────────────────────────────────
# APPROACH 4: Iterative DFS (Stack) | O(n) time | O(h) space
# EXPLAIN: Explicit stack replaces the call stack; swap children as each node is popped.
# WHEN: Avoids Python recursion limit; DFS order rather than level order.

def invertTree_stack(root: Optional[TreeNode]) -> Optional[TreeNode]:
    if not root:
        return None
    stack = [root]
    while stack:
        node = stack.pop()
        node.left, node.right = node.right, node.left
        if node.left:
            stack.append(node.left)
        if node.right:
            stack.append(node.right)
    return root


# ─────────────────────────────────────────────
# APPROACH 5: Morris Traversal | O(n) time | O(1) space
# EXPLAIN: Threaded-tree traversal; swap children at each visited node without extra memory.
# WHEN: Optimal space — O(1) extra; complex implementation, rarely required.

def invertTree_morris(root: Optional[TreeNode]) -> Optional[TreeNode]:
    curr = root
    while curr:
        if not curr.left:
            curr.left, curr.right = curr.right, curr.left
            curr = curr.left  # move to what was originally right
        else:
            pred = curr.left
            while pred.right and pred.right is not curr:
                pred = pred.right
            if not pred.right:
                pred.right = curr
                curr = curr.left
            else:
                pred.right = None
                curr.left, curr.right = curr.right, curr.left
                curr = curr.left
    return root


# Made with Bob
